Replace the test suite by the sharness test suite

roddhjav opened the pull request:
This PR replace the existing handmade tomb test suite by the [sharness](https://github.com/chriscool/sharness) test suite. See #259. It provides the portage of all the tests from the old test suite to the new one.

## Changes description

**The new `extras/test` architecture is the following:**
* `sharness`, `aggregate-results`, `Makefile` The test suite
* `setup` Internal test configuration file. It has to be sourced by all test-scripts.
* `*.sh` The tests scripts

**The tests scripts are the following:**
* `00_create.sh` tomb dig, forge, lock    
* `10_operations.sh` Basic operations on tomb: open, close
* `20_recipients.sh` GPG key related tests
* `30_kdf.sh` KDF related tests
* `40_regression.sh` Regression tests
* `50_integrity.sh` Integrity tests
* `55_index.sh`    index & search (TODO)
* `60_resize.sh` tomb resize   
* `65_passwd.sh` tomb passwd     
* `70_qrencode.sh` tomb engrave
* `75_hooks.sh` Bind hooks tests
* `80_steganography.sh` Steganography features
* `90_setkey.sh` Change the tomb key. *Should be the last test to run*

The file number allow us to require an order of execution. This is important because most of the tests have dependencies. Moreover a few tests might break the generated tomb file.

**The `setup` file also provides some helpful functions and test structure:**
* `test_export 'testname'` set the name of the tomb and tomb key(s) that can be used/reused for a given test. These names/paths are:
```sh
tomb="$TMP/$testname.tomb"
tomb_key="$TMP/$testname.tomb.key"
tomb_key_new="$TMP/$testname.new.tomb.key"
tomb_key_steg="$TMP/$testname.steg.tomb.key"
tomb_img="$TMP/$testname.jpg"
```
* Theses names are used in the following helpers functions: `tt_dig`, `tt_forge`, `tt_lock`, `tt_open` and `tt_close`. These functions allow us to write basic tests without taking care of the tomb file and key. For instance to create, open and close a tomb:
```sh
test_export 'test'
test_expect_success 'create open and close a tomb' '
    tt_dig -s 20 &&
    tt_forge --tomb-pwd $DUMMYPASS &&
    tt_lock --tomb-pwd $DUMMYPASS &&
    tt_open &&
    tt_close
    '
```

* Otherwise use `tt` to call tomb directly.

## Minors comments

* The test suite now runs fine on Archlinux (fixed regression & /media issues)
* This PR is not finalized yet, some tests are still missing.
* In order to have similar output, the verbose mode is activated in the makefile, it can be switched off.
* To run the test you can run:
  - `make test` to run all the tests
  - `zsh <test.sh>  -v` to run a file

**As usual feedbacks are welcome**

jaromil commented:
Looks great! tests pass also on the reference build on travis https://travis-ci.org/dyne/Tomb/jobs/227761258 

what are the tests still missing that you mentioned?

roddhjav commented:
What do you mean by reference build, `runtest`?. `runtest` is present (for now) but not executed by travis because `make test` runs the new test suite. That said I don't change tomb itself therefore the old test suite still works.

The remaining tests to add are: 
* `55_index.sh` that doesn't seem to be tested at all.
* `75_hooks.sh`

This is sad there is no way to have test coverage for zsh because I'm certain a lot of functions/features/conditions are not tested.

jaromil commented:
Sorry by ref. build I meant the travis platform (ubuntu 14). yea pity for the coverage really, I haven't found anything at all for zsh.

roddhjav commented:
Tests on ref build pass because I tested it on [Travis](https://travis-ci.org/roddhjav/Tomb) before to submit the PR ;).

I added bind hooks test (that now also works on an archlinux system) and I modified the Travis setup in order to compile and install the KDF

mesbahamin commented:
As someone who once contributed a test to this project, I can say that these tests are much clearer and easier to find/read/understand. I bet they will make it easier for people (especially newcomers to the project) to contribute.

jaromil commented:
many thanks for another contribution of great quality

Narrat commented:
Not sure if everything is working in the /run/media case. The permissions are kinda messed up, as all is owned by root wheras the $USER folder is set to $USER
And I ended with still occupied loop mounts (will look to pinpoint it down).

But the tests finished successfully :)

jaromil commented:
The permission errors were there also before this test suite. In any case we'll have a bit more thinkering and testing around the permission issue which also affects Tomb's production usage.
